### PR TITLE
Add groupby scan aggregation to cudf

### LINKED ISSUE #1296: [FEA] Groupby cumulative count
**Is your feature request related to a problem? Please describe.**
As a cuDF user, I want to assign numbers to each observation of a group reflecting its order of occurrence in the group.

The equivalent in the pandas API doc is [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.GroupBy.cumcount.html#pandas-core-groupby-groupby-cumcount).

**Describe the solution you'd like**
I'd like to be able to call `df.groupby(col).cumcount()` and return a column of the same size containing the numberings described above.

### LINKED ISSUE #1298: [FEA] Groupby cumulative sum
**Is your feature request related to a problem? Please describe.**
As a user, I'd like to get the cumulative sum of values within each group of a grouped dataframe.

**Describe the solution you'd like**
I'd like to call `df.groupby(col).cumsum()` and return a column of the same size containing the cumulative sum within each group.

**Describe alternatives you've considered**
The alternative to this is messy, requiring me to keep a running tally for each group as well as check the group identity of each value every time.

**Additional context**
This is blocked by #1269 .

### PR DESCRIPTION
**Pull request by karthikeyann**
closes #1296 Groupby cumulative count 
closes #1298 Groupby cumulative sum 

- [x] Add cython code for groupby scan (cannot mix reduce aggs and scan aggs)
- [x] Add python code for groupby scan functions - cumsum, cummin, cummax, cumcount, groupby.agg()
- [x] unit tests

### CODE REVIEW

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [inline python/cudf/cudf/_lib/groupby.pyx:253]: Not allowing this in cudf because pandas itself produces wrong output when both scan and reduction aggregations are passed together. (mapping of index gets wrong. could be bug in pandas)
```python
In [1]: import pandas as pd


In [2]: pdf = pd.DataFrame({'a' : [2, 2, 1, 1, 6], 'b' : [2, 2, 1, 1, 3]})


In [3]: pdf.groupby('a').agg(['sum'])
Out[3]: 
    b
  sum
a    
1   2
2   4
6   3

In [4]: pdf.groupby('a').agg(['cumsum'])
Out[4]: 
       b
  cumsum
0      2
1      4
2      1
3      2
4      3

In [5]: pdf.groupby('a').agg(['cumsum', 'sum'])
Out[5]: 
       b     
  cumsum  sum
0    2.0  NaN
1    4.0  2.0
2    1.0  4.0
3    2.0  NaN
4    3.0  NaN
6    NaN  3.0
```

Comment (shwina) [reply python/cudf/cudf/_lib/groupby.pyx:253]: 👍

Comment (kkraus14) [reply python/cudf/cudf/_lib/groupby.pyx:253]: I think we should raise an issue on Pandas regarding this because this behavior is weird.

Comment (karthikeyann) [reply python/cudf/cudf/_lib/groupby.pyx:253]: created https://github.com/pandas-dev/pandas/issues/40695

File: python/cudf/cudf/tests/test_groupby.py
Comment (kkraus14) [inline python/cudf/cudf/tests/test_groupby.py:1578]: Is this something that we should be matching the behavior of Pandas on?

Comment (karthikeyann) [reply python/cudf/cudf/tests/test_groupby.py:1578]: pandas gives `cumcount` for entire groupby object. Not on individual columns.

**Comment by kkraus14** [comment]
Pushed this to 0.20.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (kkraus14) [inline python/cudf/cudf/_lib/groupby.pyx:51]: Not the right place, but can we run `cumcount` on decimal columns as well?

Comment (karthikeyann) [reply python/cudf/cudf/_lib/groupby.pyx:51]: Yes. 

https://github.com/rapidsai/cudf/pull/7759#discussion_r605407868
> pandas gives `cumcount` for entire groupby object. Not on individual columns.

File: python/cudf/cudf/tests/test_groupby.py
Comment (karthikeyann) [reply python/cudf/cudf/tests/test_groupby.py:1605]: should this be `groupby::count()` instead?

File: python/cudf/cudf/_lib/groupby.pyx
Comment (kkraus14) [reply python/cudf/cudf/_lib/groupby.pyx:51]: I think we need to add `cumcount` to the `_DECIMAL_AGGS` dict then as well

File: python/cudf/cudf/tests/test_groupby.py
Comment (shwina) [reply python/cudf/cudf/tests/test_groupby.py:1605]: There are a few cases to consider here:

1. `df.groupby(...).cumcount()`
2. `df.groupby(...).agg('cumcount')`
3.  `df.groupby(...).agg(['cumcount'])`

Pandas returns a Series for (1) and (2), and a `DataFrame` for (3). We currently return a `DataFrame` in all cases.

(1) is easy for us to do: we define a `cumcount()` method and make sure we return a `Series`. (2) is less trivial for us:  one of the first things we do is normalize the argument passed to `agg()` into a dictionary. Thus, we lose information about whether the user passed a string, a list or a dictionary to `agg()`. 

Handling (2) is also not necessarily relevant to this PR: we have the exact same problem for the `size` aggreegation as well.

My suggestion is let's make sure we handle (1) correctly for now, and not worry about (2) in this PR.

Comment (kkraus14) [reply python/cudf/cudf/tests/test_groupby.py:1605]: Can you raise an issue about (2)?

Agreed lets tackle it in a follow up

**Review by kkraus14 — APPROVED** [review]

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [inline python/cudf/cudf/_lib/groupby.pyx:271]: @shwina 
in `_drop_unsupported_aggs`,
```python
            for i, agg_name in enumerate(aggs[col_name]):
                if Aggregation(agg_name).kind not in _DECIMAL_AGGS:
                    del result[col_name][i]
```
This logic may cause `IndexError: list assignment index out of range`.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [inline python/cudf/cudf/_lib/groupby.pyx:277]: @shwina `_drop_unsupported_aggs` may be producing wrong result.
```python

In [1]: D = {'col1': ["sum", "cumsum", "min", "max", "var", "mean", "cumcount"]}

In [2]: result = D.copy()

In [3]: AGGS = {"min", "max", "cumcount"}

In [4]: for i, val in enumerate(D['col1']):
   ...:     if val not in AGGS:
   ...:         print(i, " val=", val)
   ...:         del result['col1'][i]
   ...: 
0  val= sum
3  val= var


In [5]: result
Out[5]: {'col1': ['cumsum', 'min', 'max', 'mean', 'cumcount']}
```

Using a list comprehension should be suitable here.

What are your suggestions?

Comment (shwina) [reply python/cudf/cudf/_lib/groupby.pyx:277]: I believe https://github.com/rapidsai/cudf/pull/7818 removes this method entirely. I think at this point it would help to build on top of that PR since there will be conflicts eventually.

Comment (karthikeyann) [reply python/cudf/cudf/_lib/groupby.pyx]: Merged #7818.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/groupby.pyx:116]:
```suggestion
        cdef bool scan = _is_all_scan_aggregate(aggregations)
        if scan:
            # TODO: Once https://github.com/pandas-dev/pandas/issues/40695 is resolved, this will
            # have to also keep track of which requested aggregations are scans so that we can call
            # scan and aggregate accordingly.
            aggregations = {col_name: list(map(lambda agg_name: agg_name.replace('cum', ''), aggs))
                                       for col_name, aggs in aggregations.items()}
```

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/groupby.pyx:35]:
```suggestion
_CATEGORICAL_AGGS = {"COUNT", "SIZE", "NUNIQUE", "UNIQUE"}
```

File: python/cudf/cudf/_lib/aggregation.pyx
Comment (vyasr) [inline python/cudf/cudf/_lib/aggregation.pyx:256]: See my suggestion in groupby.pyx. I think we can just reuse the non-cumulative versions of these functions and remove all of these classmethods entirely since they're identical at the level of the creation of the C++ aggregation. This will map more closely to the C++ code since it retains a 1-1 mapping from cudf's `AggregationKind` to libcudf's `aggregation.Kind`.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (vyasr) [reply python/cudf/cudf/_lib/groupby.pyx:213]: @karthikeyann once the pandas issue is resolved and we enable mixed scan/aggregate operations, we will need to partition the creation of aggregation requests into separate lists of scan and aggregate requests so that we can call the appropriate `libcudf` function on each subset, correct? If so, would it be simpler to engineer that now, or would you prefer to wait until we reenable this? I think it's fine to wait, but assuming I understand correctly it won't be a completely trivial change since we'll have to split up the aggregations into scan and aggregate operations, farm out to `libcudf`, then recombine them in the original order, so it would be helpful to make a comment describing that somewhere.

File: python/cudf/cudf/tests/test_groupby.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_groupby.py:1639]: @shwina did you already write up the issue? If so, this conversation can be resolved.

**Review by vyasr — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing this, was waiting until the changes from #7818 could be merged in. I think we can simplify this code a little and make it match the C++ internals by mapping cumulative operations to their non-cumulative counterparts earlier, but I've left a couple of comments to ascertain that this won't cause significant problems when we start supporting mixed scan/aggregate operations.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [reply python/cudf/cudf/_lib/groupby.pyx:213]: pandas issue should be resolved and decide the behavior. Right now, it seems, not supporting this is good idea otherwise this might become a breaking change for cudf. Based on the behavior decided, the change could be trivial or not.

File: python/cudf/cudf/_lib/groupby.pyx
Comment (karthikeyann) [reply python/cudf/cudf/_lib/groupby.pyx:116]: Not all aggregations are strings. There may be functions too.

File: python/cudf/cudf/_lib/aggregation.pyx
Comment (karthikeyann) [reply python/cudf/cudf/_lib/aggregation.pyx:256]: Right now, they are identical. When https://github.com/rapidsai/cudf/issues/7106 is implemented, they will not be identical. Hence separate functions.
Besides, `cumcount` is different.

One suggestion to avoid repetition could be to create alias.
```python
    cumsum = sum
    cummin = min
    cummax = max
```

Comment (vyasr) [reply python/cudf/cudf/_lib/aggregation.pyx:256]: Yes I like that solution. I agree that this will change when the C++ side is modified to be stricter about providing the types available, but until then I think it's helpful to match the C++ as closely as possible as well as avoiding duplication here if we can.

**Review by vyasr — APPROVED** [review]
One or two minor comments from me, otherwise looks ready.

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**
